google/cloud/internal/async_retry_loop.h

```diff
@@ -93,120 +197,120 @@ class AsyncRetryLoopImpl
+               cq_.MakeRelativeTimer(backoff_policy_->OnCompletion())
+                   .then([self](future<TimerArgType> f) {
+                     self->OnBackoff(f.get());
+                   }));
   }
 
   void OnAttempt(T result) {
-    SetIdle();
     // A successful attempt, set the value and finish the loop.
-    if (result.ok()) {
-      SetDone(std::move(result));
-      return;
-    }
+    if (result.ok()) return SetDone(std::move(result));
     // Some kind of failure, first verify that it is retryable.
     last_status_ = GetResultStatus(std::move(result));
     if (idempotency_ == Idempotency::kNonIdempotent) {
-      SetDone(RetryLoopError("Error in non-idempotent operation", location_,
-                             last_status_));
-      return;
+      return SetDone(RetryLoopError("Error in non-idempotent operation",
+                                    location_, last_status_));
     }
     if (!retry_policy_->OnFailure(last_status_)) {
       if (retry_policy_->IsPermanentFailure(last_status_)) {
-        SetDone(RetryLoopError("Permanent error in", location_, last_status_));
-      } else {
-        SetDone(RetryLoopError("Retry policy exhausted in", location_,
-                               last_status_));
+        return SetDone(
+            RetryLoopError("Permanent error in", location_, last_status_));
       }
-      return;
+      return SetDone(
+          RetryLoopError("Retry policy exhausted in", location_, last_status_));
     }
-    if (Cancelled()) return;
-    auto self = this->shared_from_this();
-    auto op =
-        cq_.MakeRelativeTimer(backoff_policy_->OnCompletion())
-            .then(
-                [self](
-                    future<StatusOr<std::chrono::system_clock::time_point>> f) {
-                  self->OnBackoffTimer(f.get());
-                });
-    SetWaiting(std::move(op));
+    StartBackoff();
   }
 
-  void OnBackoffTimer(StatusOr<std::chrono::system_clock::time_point> tp) {
-    SetIdle();
-    if (Cancelled()) return;
+  void OnBackoff(TimerArgType tp) {
+    auto state = OnOperation();
+    // Check for the retry loop cancellation first. We want to report that
+    // status instead of the timer failure in that case.
+    if (state.cancelled) return;
     if (!tp) {
       // Some kind of error in the CompletionQueue, probably shutting down.
-      SetDone(RetryLoopError("Timer failure in", location_,
-                             std::move(tp).status()));
-      return;
+      return SetDone(RetryLoopError("Timer failure in", location_,
+                                    std::move(tp).status()));
     }
     StartAttempt();
   }
 
-  void SetIdle() {
-    std::unique_lock<std::mutex> lk(mu_);
-    switch (state_) {
-      case kIdle:
-      case kDone:
-        break;
-      case kWaiting:
-        state_ = kIdle;
-        break;
-    }
-  }
-
-  void SetWaiting(future<void> op) {
+  void SetPending(std::uint_fast32_t iteration, future<void> op) {
```

Comment:
do you still want this parameter to be called `iteration`?

Reply:
Nope, fixed.